### PR TITLE
Stan integration #258

### PR DESCRIPTION
**Pull request by uhbif19**
#258

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2908"></a>

### CODE REVIEW

File: plugins/hls-stan-plugin/test/testdata/test.hs
Comment (Ailrun) [inline plugins/hls-stan-plugin/test/testdata/test.hs:4]: The test directory contains some compiler outputs, such as .o and .hi.

Comment (uhbif19) [reply plugins/hls-stan-plugin/test/testdata/test.hs:4]: Fixed. Shouldn't they be in gitignore? I can add them in this or separate PR.

Comment (Ailrun) [reply plugins/hls-stan-plugin/test/testdata/test.hs:4]: I think we (most of HLS contributors) rarely compile files directly so they aren't in .gitignore (the file contains stack's and cabal's working directories only)

File: CODEOWNERS
Comment (michaelpj) [inline CODEOWNERS:31]: Only works if you've got write access... I've invited you :)

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (michaelpj) [inline plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:74]: Is this idea to make this configurable? Comment?

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (michaelpj) [inline plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:99]: Please don't use partial functions, this can take down HLS entirely. You can log an error and just omit the case you can't handle, just don't `error`.

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (michaelpj) [inline plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:100]:
```suggestion
        -- Looking similar to Stan CLI output
```

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (michaelpj) [inline plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:110]: This seems like something we should provide alongside the other HIE stuff? WDYT @wz1000 ?

Comment (michaelpj) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:110]: At least this needs some comments.

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (uhbif19) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:74]: We can enable Inspections by category: https://github.com/kowainik/stan/blob/main/src/Stan/Inspection/All.hs#L31.

Or we can add `excludeInspections` option.

Also Stan may look for cabal config to look for extensions. Is it possible to ask HLS for cabal file?

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (uhbif19) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:110]: @michaelpj For me this function signature looks pretty self-explanatory. What else could be said about it?

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (uhbif19) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:99]: But this function cant fail, because inspection ID always should be from inspection id list.

Comment (uhbif19) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs]: Fixed.

**Comment by uhbif19** [comment]
I do not understand why CI fails. Plugin is disabled for GHC < 9, but Cabal build still fails: https://github.com/haskell/haskell-language-server/pull/2908/files#diff-aaf5dad5ddf1d2c16a2e8034c6435719f6eb3327cfe1bcd042df5a84b939afe3R282

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (michaelpj) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:111]: Here are a list of questions that I have about this function after reading it:
- Why isn't this function provided with the other HIE file functions we have? Is it doing something suspicious? Is there a reason we don't want to provide HIE files? Is it working against the direction of travel with hiedb?
- How does it work? Are we sure we're constructing HIE files correctly? Can the information we get from the rules be inconsistent with what's coming from hiedb? 

I'm sure you thought about a bunch of these things when you were writing it, but all that knowledge is going to be lost unless it gets written down! :)

**Comment by michaelpj** [comment]
@pepeiborra changed how you have a plugin that's only enabled on some versions recently. I'm not sure what the right way is off the top of my head (@pepeiborra if you write something here I'll add it to the contributing docs).

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (uhbif19) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:111]: This function was written by @pepeiborra , so I do not know the answers https://github.com/haskell/haskell-language-server/issues/258#issuecomment-1126890590

**Comment by pepeiborra** [comment]
You need to set the `buildable` field of your library and possibly the test-suite too.Example

https://github.com/haskell/haskell-language-server/blob/8cbefb71bc3aff7c19d4874607b6a374f2307fc3/plugins/hls-haddock-comments-plugin/hls-haddock-comments-plugin.cabal#L20-L24

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (pepeiborra) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:111]: Those are valid questions, and @wz1000 is the right person to review rather than me. Also, perhaps move this function to the `ghcide` package and refactor other pieces of similar code around the `Core/Compile.hs` and `Core/Rules.hs` modules.

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (July541) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:74]: Maybe a configure file like https://github.com/kowainik/stan/blob/main/.stan.toml?

Comment (michaelpj) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:74]: I'm not saying we should necessarily make it configurable now, but it would be good to write something down here about it.

e.g.
```
-- In the future we may want to control which inspections are enabled based on ???
```

> Also Stan may look for cabal config to look for extensions.

HLS knows the extensions enabled for a file directly, if `stan` will let you set them.

Comment (uhbif19) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:74]: > HLS knows the extensions enabled for a file directly

I guess Stan waits for cabal enabled extensions, not extensions enabled directly, cuz function used for collecting extensions is named `createCabalExtensionsMap`. https://github.com/kowainik/stan/blob/main/src/Stan.hs#L74

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (Ailrun) [inline plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:89]: Is there a reason to avoid `prettyShowInspection` of `Stan.Inspection`? If so, could you expand this comment with the reason?

Comment (uhbif19) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:89]: `prettyShowInspection` included "Id" (which is already in _code) and "Severity" (which could be placed in _severety), which are not needed to show here, as I think.

Comment (michaelpj) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:89]: Worth a comment.

Comment (uhbif19) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:95]: Comment done.

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (uhbif19) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs]: Comment done.

File: haskell-language-server.cabal
Comment (July541) [inline haskell-language-server.cabal:365]: It looks you replaced hlint with stan?

**Comment by July541** [comment]
Can we have a document as #2066 mentioned?

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (michaelpj) [inline plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:96]: You're including "categories", though?

**Comment by michaelpj** [comment]
Yes, please do add something to the "Features" doc.

I would like @wz1000 to look at the hiedb stuff before we merge this.

**Comment by uhbif19** [comment]
Marked as draft, to rebase before merging.

**Comment by uhbif19** [comment]
I tried to use plugin, but it does not show anything to me.

Does something in VS Code Haskell need to be tweaked for this to work?
https://github.com/haskell/vscode-haskell/blob/e81160dbdb6be0ac8493f6dbc0983d27fa096ac7/package.json#L346

**Comment by pepeiborra** [comment]
> Does something in VS Code Haskell need to be tweaked for this to work?

Not at all. I am in my phone right now, but if the test suite passes you should see the diagnostics in VSCode already.

**Comment by michaelpj** [comment]
I'm not sure that's true. I think all plugins automatically get a "globalOn" option that needs to be set to enable them. I'm not sure if there's a way to turn it on by default when testing (@berberman ?).

For vscode, we have an ongoing issue that nobody seems to be sure what the deal is with options in the vscode extension (see https://github.com/haskell/haskell-language-server/issues/2827). It would be great if someone could find out and help us document it.

**Comment by July541** [comment]
> if the test suite passes you should see the diagnostics in VSCode already.

Agree this.

Noticed you are using hint severity,it may not showy as you expected in vscode.

**Comment by berberman** [comment]
> I'm not sure that's true. I think all plugins automatically get a "globalOn" option that needs to be set to enable them. I'm not sure if there's a way to turn it on by default when testing (@berberman ?).

If you don't specify `globalOn` in the lsp config for a plugin, which can be set by `sendNotification SWorkspaceDidChangeConfiguration DidChangeConfigurationParams ...` during the test, the default value `True` will be used:
https://github.com/haskell/haskell-language-server/blob/310e6a4683c40dc5d93ffaf12c45144dc8e079a4/hls-plugin-api/src/Ide/Plugin/Config.hs#L116-L118

So no worries about `globalOn`.

> For vscode, we have an ongoing issue that nobody seems to be sure what the deal is with options in the vscode extension (see #2827). It would be great if someone could find out and help us document it.

Yes, I can write a documentation for that later.

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (uhbif19) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:122]: @wz1000 Please, tell how to refactor this function?

**Comment by pepeiborra** [comment]
is this still in progress or ready for review?

**Comment by pepeiborra** [comment]
@uhbif19  there are stack build failures

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (pepeiborra) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs:122]: > Also, perhaps move this function to the ghcide package and refactor other pieces of similar code around the Core/Compile.hs and Core/Rules.hs modules.

Did you have a chance to look into this refactoring?

**Review by pepeiborra — APPROVED** [review]
Looks good to me, but it would be nice to make the suggested refactoring

File: .github/workflows/test.yml
Comment (uhbif19) [reply .github/workflows/test.yml:226]: Stan does not support 9.x

File: plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs
Comment (uhbif19) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs]: I thought, I was waiting on suggestion where to place the function.

Now I moved `getHieFile` to Development.IDE.Core.Rules.

As for `refactor other pieces of similar code around the Core/Compile.hs and Core/Rules.hs modules` I do not understand what you are talking about. Can you explain what you want to be performed?

Comment (pepeiborra) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs]: I believe that we have code to create HieFile in Core.Rules and/or Core.Compile

It would be nice to merge it with getHieFile in order to ensure that it stays working.

Comment (uhbif19) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs]: You mean create separate rule for HieFile?

Comment (pepeiborra) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs]: No, just abstracting and reusing the same code everywhere

Comment (uhbif19) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs]: @pepeiborra I do not know which code should be abstracted.

Comment (pepeiborra) [reply plugins/hls-stan-plugin/src/Ide/Plugin/Stan.hs]: Fair enough

**Comment by uhbif19** [comment]
@michaelpj CircleCI stuck again.

**Pull request merged**
